scilifelab_epps/epp.py

```diff
@@ -467,3 +469,89 @@ def get_well_number(art: Artifact, count_per: str) -> int:
         raise AssertionError
 
     return well_num
+
+
+def get_matching_inputs(
+    process: Process, output_artifact: Artifact
+) -> list[Artifact] | None:
+    """Get the input artifacts coupled to an output artifact."""
+    input_arts = []
+    for io_tuple in process.input_output_maps:
+        input_art = io_tuple[0]["uri"]
+        output_art = io_tuple[1]["uri"]
+        if input_art.type == "Analyte" and output_art.id == output_artifact.id:
+            input_arts.append(io_tuple[0]["uri"])
+
+    if input_arts:
+        return input_arts
+    else:
+        return None
+
+
+def traceback_to_step(
+    art: Artifact, step_name_pattern: re.Pattern, allow_multiple_inputs: bool = False
+) -> tuple[Process, list[Artifact], Artifact] | None:
+    """For an artifact try to backtrack it to a target step based on a supplied pattern.
+
+    Returns:
+    - The target step
+    - A list of it's matching input Artifacts
+    - It's matching output Artifact
+
+    If a step is reached where the artifact has multiple linked inputs, linear traceback is not possible.
```

Comment:
and if it's not possible, it will raise an exception? Maybe clarify that?

Reply:
Maybe an example of how the pattern is used could be good to help my understanding...

Reply:
c703600cb0d21debfd8e007552d08d422ab17e3e
7aa41d6a2a0b96968fe418d3bfc352deb0ea49c7